File: source/yeon/4주차/code/bucketSort.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<random>
#define LEN 8

using namespace std;

void bucketSort(float arr[], int len){
    vector<float> bucket[8];

    for(int i = 0; i < len; i++){
        int index = len * arr[i];
        bucket[index].push_back(arr[i]);
    }

    for (int i = 0; i < len; i++){
        sort(bucket[i].begin(), bucket[i].end());
    }
    
    int index = 0;
    for (int i = 0; i < len; i++){
        for (int j = 0; j < bucket[i].size(); j++){
            arr[index++] = bucket[i][j];
        }
    }
}

int main(void){
    srand(time(NULL));
    float arr[LEN] = {0.0,};
    for (int i = 0; i < LEN; i++){
        arr[i] = rand() / (float)RAND_MAX * 1.0f ;
    }
    bucketSort(arr, LEN);
    for(int i = 0; i < LEN; i++){
        printf("%f ", arr[i]);
    }
    return 0;
}

File: source/yeon/4주차/code/bucketSort.py
def insertionSort(bucket):
    for i in range(1,len(bucket)):
        min = bucket[i]
        j = i - 1
        while j >= 0 and bucket[j] > min :
            bucket[j+1] = bucket[j]
            j -= 1
        bucket[j+1] = min
    return bucket

def bucketSort(arr):
  	#10 means 10 slots, slot's size if 0.1
    bucket = [[] for _ in range(10)]
	
  	#put elements in each buckets
    for i in arr :
        index = int(i * 10)
        bucket[index].append(i)
        
		#sort individual buckets
    for i in range(10):
        bucket[i] = insertionSort(bucket[i])
    
    #concatenate the bucket
    x = 0
    for i in range(10):
        for j in range(len(bucket[i])):
            arr[x] = bucket[i][j]
            x += 1
    
    
arr = [0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434, 0.12, 0.24] 
print("정렬 전 : ", arr, end=" ") 
bucketSort(arr)
print ("-> 정렬 후 : ",arr) 